GPSUARTDriver.device.lib.nut

```diff
@@ -0,0 +1,110 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+// EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+// Driver for UART GPS modules. This driver is dependant on the GPSParser class.
+// This driver is focused on retrieving location data and returns data from the
+// GPSParser.getGPSDataTable method. Currently only the data with sentence IDs
+// VTG, RMC, GLL, GGA, GSV and GSA are supported.
+class GPSUARTDriver {
+
+    function _statics_() {
+        const VERSION          = "1.0.0";
+        const LINE_MAX         = 150;
+        const DEFAULT_BAUDRATE = 9600;
+    }
+
+    _gps            = null;
+    _gpsLine        = null;
+    _callback       = null;
+
+    _hasFix         = null;
+    _lastLat        = null;
+    _lastLong       = null;
+
+    constructor(uart, baudrate = null, dataReady = null) {
+        _gps     = uart;
+        _hasFix  = false;
+        _gpsLine = "";
+
+        // Check optional params
+        if (typeof baudrate == "function") {
+            _callback = baudrate;
+            baudrate = DEFAULT_BAUDRATE;
+        } else if (baudrate == null) {
+            baudrate = DEFAULT_BAUDRATE;
+            _callback = dataReady;
+        } else {
+            _callback = dataReady;
+        }
+
+        // GPS is configured by the constructor so that we can register the rxdata callback
+        _gps.configure(baudrate, 8, PARITY_NONE, 1, NO_CTSRTS, _gpsRxData.bindenv(this));
+    }
+
+    function hasFix() {
+        return _hasFix;
+    }
+
+    function getLatitude() {
+        return _lastLat;
+    }
+
+    function getLongitude() {
+        return _lastLong;
+    }
+
+    // This private method is the uart callback. It continues to append characters
+    // to a line until it reaches a '$', indicating the start of a new line. Once it reaches this,
+    // it parses the previous line and calls callbacks (if provided), as well as setting
+    // values in the class that can be accessed (e.g. lat and long)
+    function _gpsRxData() {
+        local ch = _gps.read();
```

Comment:
I left this alone since it seemed to be working, but I think I'll rewrite this function entirely. I think we should be looking for the last character in the sentence not the first. This chip spits out data constantly, so you can wait for the next message to start before parsing the previous data, but I don't think this is a good idea in general.